feat: add static orbit

ngwanzhen opened the pull request:
Static orbit for now. Readme and dynamic orbit to follow.

Commit:
feat: add static orbit

ghost commented:
## DeepCode's analysis on [#5f8041](https://www.deepcode.ai/app/gh/matteobruni/tsparticles/2c8d040a15b2c3f5db1881bad6eb25cca9cf9c81/matteobruni/tsparticles/5f8041847ad1c43f0b216c8fead3ab385369f887/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=1&) found:

- :information_source: **1** minor issue. :point_down:
## Top issues

<table>
  <thead>
    <tr>
      <th align="left">Description</th>
      <th align="left">Example fixes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td width="77%">'public' is implicit. Occurrences: <ul><li><a href="https://github.com/matteobruni/tsparticles/blob/5f8041847ad1c43f0b216c8fead3ab385369f887/core/main/src/Options/Classes/Particles/Particles.ts#L33">Particles.ts:33</a></li> <li><a href="https://github.com/matteobruni/tsparticles/blob/5f8041847ad1c43f0b216c8fead3ab385369f887/core/main/src/Options/Classes/Particles/Particles.ts#L42">Particles.ts:42</a></li> <li><a href="https://github.com/matteobruni/tsparticles/blob/5f8041847ad1c43f0b216c8fead3ab385369f887/core/main/src/Options/Classes/Particles/Particles.ts#L50">Particles.ts:50</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/matteobruni/tsparticles/2c8d040a15b2c3f5db1881bad6eb25cca9cf9c81/matteobruni/tsparticles/5f8041847ad1c43f0b216c8fead3ab385369f887/pr/_/%2Fcore%2Fmain%2Fsrc%2FOptions%2FClasses%2FParticles%2FParticles.ts/javascript%2Ftslint%2Fmember-access/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
  </tbody>
</table>


#### 👉 View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/matteobruni/tsparticles/2c8d040a15b2c3f5db1881bad6eb25cca9cf9c81/matteobruni/tsparticles/5f8041847ad1c43f0b216c8fead3ab385369f887/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=1&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=matteobruni)

matteobruni commented on core/main/src/Core/Particle.ts:
```diff
@@ -106,6 +108,7 @@ export class Particle implements IParticle {
 
         const shapeType = particlesOptions.shape.type;
         const reduceDuplicates = particlesOptions.reduceDuplicates;
+        this.orbitOptions = particlesOptions.orbit;
```
This should not be done, the object is referenced instead of assigning the needed values. This can cause issues when reloading options or changing values manually.

ngwanzhen replied:
Removed.

matteobruni replied:
More details [here](https://github.com/matteobruni/tsparticles/pull/986#discussion_r513235768)

matteobruni commented on core/main/src/Core/Canvas.ts:
```diff
@@ -489,6 +490,22 @@ export class Canvas {
+            this.context,
+            particle,
+            orbitOptions.color || particle.getFillColor(),
+            orbitOptions.radius || particle.getRadius(),
```
Looking at code this radius seems just the value taken directly from the config, but it needs to be updated with retina support like all other radius properties and distances.

ngwanzhen replied:
I've added a method in Retina to fix radius, not sure what to do about distances? Would be great if you could elaborate more on that.
`public initParticleOrbit(particle: Particle): void {
        const ratio = this.pixelRatio;
        particle.particlesOptions.orbit.radius = particle.particlesOptions.orbit.radius * ratio;
    }
`

matteobruni replied:
A new method is not necessary. You can add it to the `initParticle` method.

I was talking about what could be affected by retina displays, if you are not using distances ignore that part. Checkout how other radiuses work.

Like I said in another comment the `particlesOptions` object should be only read, and the orbit options should be stored inside the `Particle` object like many other values.

I want `options` to be read-only because they are just input values selected by the user, changing them can bring unexpected results when reloading.

matteobruni commented on core/main/src/Core/Particle.ts:
```diff
@@ -408,6 +411,10 @@ export class Particle implements IParticle {
 
     public draw(delta: IDelta): void {
+        if (this.orbitOptions.enable) {
+            this.container.canvas.drawOrbit(this, this.orbitOptions);
```
This should be inside the `drawParticle` method

ngwanzhen replied:
Moved to drawParticle.

matteobruni commented on core/main/src/Options/Interfaces/Particles/IParticles.ts:
```diff
@@ -45,4 +46,5 @@ export interface IParticles {
     size: ISize;
     stroke: SingleOrMultiple<IStroke>;
     twinkle: ITwinkle;
+    orbit: IOrbit;
```
Can you please sort it alphabetically? It's easier to update docs

ngwanzhen replied:
Sorted.

matteobruni replied:
I still see it under `twinkle`, but it should be under `opacity`. `@deprecated` properties are ignored in the sort, they won't be documented.

Commit:
refactor: refactor orbit code

Commit:
refactor: refactor orbit code

matteobruni reviewed (changes requested):
There's still something to review, more comments to the existing conversations

ngwanzhen commented on core/main/src/Options/Interfaces/Particles/IParticles.ts:
```diff
@@ -38,6 +39,7 @@ export interface IParticles {
     move: IMove;
     number: IParticlesNumber;
     opacity: IOpacity;
+    orbit: IOrbit;
```
I see it under opacity? strange? whats missing here... @matteobruni

matteobruni replied:
Ok now it's fine

Commit:
refactor: refactor orbit code

matteobruni commented:
Now everything looks good, can you create a sample in `demo/main` to see it working?

Commit:
feat: add orbit to demo

matteobruni reviewed (changes requested):
The `rotate` configuration is not working, the orbit doesn't rotate with the particle

matteobruni commented:
I've seen some improvements that can be done:

- `width` and `radius` are set to `0` but orbits are visible. This shouldn't happen. You can also change default values to optimal values, just keep them disabled by default.
- `rotation` angle at `0` should be horizontal, you can default the value to 45 degrees if you prefer that value for default. The angle in options must be in degrees and converted to `rad` when rotating it in code.
- `rotation`, `radius`, `opacity`, `color`, and `width` could have `random` and `animation` properties to have different values for each particle and animations set to each one of them (rotating radius with a non-rotating particle, orbit changing size in time bouncing from max to min, orbit appearing/disappearing, etc.). (You can see how particles `opacity`, `color`, and `size` work)

Commit:
fix: orbit width radius rotation

ngwanzhen commented:
> I've seen some improvements that can be done:
> 
> * `width` and `radius` are set to `0` but orbits are visible. This shouldn't happen. You can also change default values to optimal values, just keep them disabled by default.
> * `rotation` angle at `0` should be horizontal, you can default the value to 45 degrees if you prefer that value for default. The angle in options must be in degrees and converted to `rad` when rotating it in code.
> * `rotation`, `radius`, `opacity`, `color`, and `width` could have `random` and `animation` properties to have different values for each particle and animations set to each one of them (rotating radius with a non-rotating particle, orbit changing size in time bouncing from max to min, orbit appearing/disappearing, etc.). (You can see how particles `opacity`, `color`, and `size` work)

Ok, I've fixed `width`, `radius`, `rotation`. (FYI, lineWidth in canvas is defaulted to 1 by default. 0 and NaN are ignored. [https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineWidth] So I've worked around it by not drawing if width input is 0).

 As for `random` and `animation`, those are features I would also like to incorporate. But is it possible to merge this first, and open new issues so that me / others can come back to it at a later time?

Commit:
refactor: orbit width

Commit:
refactor: orbit width

Commit:
refactor: orbit

matteobruni commented on core/main/src/Core/Interfaces/IOrbit.ts:
```diff
@@ -0,0 +1,13 @@
+    enable: boolean;
+    opacity: number;
+    width: number;
+    color: IHsl | undefined;
```
Use the optional syntax for optional propperties instead of ` | undefined`

```javascript
color?: IHsl;
```

matteobruni commented on core/main/src/Core/Interfaces/IParticle.ts:
```diff
@@ -47,6 +47,7 @@ export interface IParticle {
     readonly moveSpeed?: number;
     readonly sizeValue?: number;
     readonly sizeAnimationSpeed?: number;
+    readonly orbitRadiusValue?: number | undefined;
```
`| undefined` is not necessary since the property is already optional

matteobruni commented on core/main/src/Core/Particle.ts:
```diff
@@ -63,6 +63,7 @@ export class Particle implements IParticle {
     public moveSpeed?: number;
     public sizeValue?: number;
     public sizeAnimationSpeed?: number;
+    public orbitRadiusValue?: number | undefined;
```
`| undefined` is not necessary since the property is already optional

matteobruni commented on core/main/src/Utils/CanvasUtils.ts:
```diff
@@ -460,4 +460,33 @@ export class CanvasUtils {
+
+        context.lineWidth = width;
+
+        const rotationRadian = (rotation + 90) * (Math.PI / 180);
```
Why `rotation + 90` instead of `rotation`?

ngwanzhen replied:
The default behaviour at 0 is a vertical ellipse instead of horizontal, like you asked.

matteobruni replied:
Oh ok, My bad I was misunderstood. The `rotation` value should be represented as is set. If `0` is set, a horizontal line should be drawn. In the preset that you created for testing the feature the value was `0` but the orbit was drawn with a 45° angle.

ngwanzhen replied:
The current behaviour is that `rotation` has a default value of `45` and will show as `45` in the editor. 
If you change that value to 0, a horizontal ellipse will be drawn. 
If I remove the `+90`, default behaviour will still be 45. But at 0, a vertical ellipse will be drawn instead. And at 90, a horizontal one will be drawn. Should I remove?

matteobruni replied:
Yes remove the 90, I've seen 0 and a 45-degree orbit in the test I did some commits ago, probably it's fine now. Once fixed, I'll test it again.

matteobruni commented on core/main/src/Core/Canvas.ts:
```diff
@@ -487,6 +488,27 @@ export class Canvas {
+            this.context,
+            particle,
+            orbitOptions.color || particle.getFillColor(),
+            typeof particle.orbitRadiusValue === "undefined" ? particle.getRadius() : particle.orbitRadiusValue,
```
you can shorten this with the null check:

```javascript
particle.orbitRadiusValue ?? particle.getRadius()
```

ngwanzhen replied:
This was the previous code, but when setting radius as 0, you will end up not using 0 and use the particle's radius instead as its falsey I believe.

matteobruni replied:
Probably you were using the `||` operator which uses all falsy values, the `??` operator works only with `undefined` and `null`.

ngwanzhen replied:
Ok, that's cool. Didn't know that. thanks!

matteobruni commented:
> > I've seen some improvements that can be done:
> > 
> > * `width` and `radius` are set to `0` but orbits are visible. This shouldn't happen. You can also change default values to optimal values, just keep them disabled by default.
> > * `rotation` angle at `0` should be horizontal, you can default the value to 45 degrees if you prefer that value for default. The angle in options must be in degrees and converted to `rad` when rotating it in code.
> > * `rotation`, `radius`, `opacity`, `color`, and `width` could have `random` and `animation` properties to have different values for each particle and animations set to each one of them (rotating radius with a non-rotating particle, orbit changing size in time bouncing from max to min, orbit appearing/disappearing, etc.). (You can see how particles `opacity`, `color`, and `size` work)
> 
> Ok, I've fixed `width`, `radius`, `rotation`. (FYI, lineWidth in canvas is defaulted to 1 by default. 0 and NaN are ignored. [https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineWidth] So I've worked around it by not drawing if width input is 0).
> 
> As for `random` and `animation`, those are features I would also like to incorporate. But is it possible to merge this first, and open new issues so that me / others can come back to it at a later time?

I'd prefer to merge it when the feature is completed. This will be in `1.19.0` release and for now, I can wait.

I need to work on that version too and probably will be released at least at the end of November.

Commit:
refactor: orbit

Commit:
refactor: orbit

Commit:
refactor: orbit

matteobruni commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/main/src/Core/Canvas.ts  1
- core/main/src/Options/Classes/Particles/Orbit/Orbit.ts  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/matteobruni/tsparticles/pullRequest?prid=6406861&bid=20766004)

matteobruni commented:
Can you change the orbit drawing like this: https://codepen.io/matteobruni/pen/YzWeRXe?editors=0010

There are 3 samples, one with the particle drawn after the orbit, one before, and the last one (the best solution) is drawing the back orbit, the particle, and the front orbit.

ngwanzhen commented:
> Can you change the orbit drawing like this: https://codepen.io/matteobruni/pen/YzWeRXe?editors=0010
> 
> There are 3 samples, one with the particle drawn after the orbit, one before, and the last one (the best solution) is drawing the back orbit, the particle, and the front orbit.

This looks really good! Thanks for the very clear explanation! I'll add it, but maybe next week.

matteobruni committed:
Merge branch 'dev' into 188-add-orbit-feature

matteobruni committed:
Merge branch 'dev' into 188-add-orbit-feature

Commit:
feat: animation, random for orbit angle and redrawn orbit

Commit:
Merge branch '188-add-orbit-feature' of https://github.com/thecodingdog/tsparticles into 188-add-orbit-feature

Commit:
refactor: orbit

matteobruni commented on core/main/src/Core/Canvas.ts:
```diff
@@ -434,8 +435,26 @@ export class Canvas {
+        const orbitOptions = particle.particlesOptions.orbit;
+        const zSizeFactor = 1 - zIndexOptions.sizeRate * particle.zIndexFactor;
+
+        if (radius > 0 && orbitOptions.enable === false) {
```
The `drawParticle` code is duplicated and long, using two `if`s to check if the orbits need to be drawn is shorter and safer, and an enum instead of `"front"` and `"back"` should be a better solution

matteobruni reviewed (changes requested):
Just the small change to keep the code cleaner and we're done, good job!

Commit:
refactor: orbit
